### PR TITLE
Add support for applying effects on future completion

### PR DESCRIPTION
**Pull request by odd**
This allows the following kind of code:
```scala
case Increment =>
  import FutureEffect._
  implicit val ec: ExecutionContext = ...
  val f: Future[Int] = fetchHowMuchToIncrement()
  ctx.effectOnComplete(f) {
    case Success(n) => Effect.persist(Incremented(n))
    case Failure(t) => Effect.none
  }
``` 
There is no need for the user to define any internal commands to handle the callback, instead an Akka internal `ApplyEffect` message captures the effect generated in the user supplied callback. The effect will then be applied automatically by Akka as part of the pre command interception handling.

### CODE REVIEW

**Comment by odd** [comment]
Both the **Jenkins PR Validation** check and the **continuous-integration/travis-ci/pr** check seem to have succeeded but somehow GitHub hasn't noticed. Can the GitHub checks check be made to forcible recheck somehow?

**Comment by johanandren** [comment]
They don't actually run until we OK them.

I vaguely remember that when we discussed adding this but decided not to because it makes it unclear what happens in the meanwhile, that we'd rather see that become an ask and a new command with the needed dependent data or async validation results.

**Comment by odd** [comment]
> They don't actually run until we OK them.

Thanks, that explains it! 😄 

> I vaguely remember that when we discussed adding this but decided not to because it makes it unclear what happens in the meanwhile, that we'd rather see that become an ask and a new command with the needed dependent data or async validation results.

I don't fully understand what "what happens in the meanwhile" refers to, could you please elaborate. 

I've tried to summarize the pros/cons for adding this feature below:

**Pros**
* No need to define actor specific internal commands just to pass results back to the actor making integrating event sourced actors with asynchronous services returning futures much easier (the result/effect carrying messages and handling of those are instead internal to Akka).
* The effect producing logic can be located close to the origin of the asynchronous future result (i.e. it can be located under the same command handling case as the originating command instead of having to move it to its own command handling case block).

**Cons**
* Potential confusion as to when (if ever) the resulting effect will be persisted (although the same can be said of the current method of sending explicit result carrying commands back to the actor).
* A bit clunky syntax wise (I tried to make it as lightweight as I could but that might be possible to improve).
* Non obvious method name (I did try some alternatives such as `applyOnComplete` and `async` but didn't find anything smashing).

**Comment by johanandren** [comment]
The actor accept and validates the command against current state, then trigger the async task that completes at some future point in time, while the future is in flight it may accept commands leading to non-async persists, when the future completes it may no longer be valid to persist the event - the original command was validated against a state that is no longer valid.

I think being explicit in what happens while the async dependency is resolved and also how/if you handle failures/timeouts is worth the boilerplate as it makes it explicitly clear what happens in between accepting the first command and the second (in your case implicit) command.

**Comment by odd** [comment]
Thanks for the explanation. If I understand you correctly it is the fact that the effect is decided while handling the original command and then applying that effect at a later time (by which time the original decision could be bad). That is indeed a relevant objection to transferring the resulting effect. 

Perhaps by instead transferring a function from `State` to `Effect[Event, State]` that problem could be avoided while still having the benefit of reducing the plumbing boilerplate associated with a user level command? To avoid this boilerplate (which needs to be repeated for each command hierarchy) is the ultimate goal of this PR. 

The Akka internal command would then look like:
```scala
final case class ApplyEffect[E, S](f: S => Effect[E, S]) extends InternalProtocol
```
and the usage something like:
```scala
case Increment =>
  import FutureEffect._
  implicit val ec: ExecutionContext = ...
  val f: Future[Int] = fetchHowMuchToIncrement()
  ctx.effectOnComplete(f) {
    case Success(n) => state => Effect.persist(Incremented(...state based logic...))
    case Failure(t) => _ => Effect.none
  }
```

I'll update this PR with this change.

**Comment by renatocaval** [comment]
I find it a strange feature to add. The internal state of an EventSourcedBehavior should not be determined by an external call. 

Event-sourcing is about modelling a consistency boundary and that means that any command should be validated against the data directly available in scope. If in order to make such a decision, we need to fetch data on some other place, then the model can not be guaranteed to be consistent. Unless we add extra APIs for that, but I don't see the need for that. 

If we need to go fetch some extra data, we should instead do it before sending the command to the actor. I mean, the command should be enriched with the remote data before it's sent to the actor. Then the actor can validate it. 

The only use case that comes to my mind that would justify it is when you need to use the actor data to call the remote service. In that case, you need to be inside the actor. But then I will argue that the use case is so limited that we should not extend the API to support it. 

It is already possible to do it. One can always call a remote service and send a command to itself when it completes. And we can choose if we switch to another state or not (with or without stashing) while we wait for the remote completion. 

To summarize my point: 

1. validation should be pure in the sense that it should have all data in scope. Either in the command or in the actor state. 
1. if needed, commands should be enriched before they are sent to actor
1. the rare case where actor data is needed to call remote service can already be implemented using existing APIs

**Comment by renatocaval** [comment]
This is another issue we have in which we discussed the idea of async effects:
https://github.com/akka/akka/issues/25650

The context is a little bit different. It's not about validating the commands with external data, but an interaction pattern between an Actor and a Service (or another Actor). 

1. actor receives command, process and calls external service
1. when the remote call completes, actor move to next action

The conclusion was that a completed Future should be a new command sent to self.

**Comment by odd** [comment]
Hi Renato, and thanks for evaluating this PR.

Just to clarify, the intended use case is very similar to the one you describe in your last comment, although in this case the next action to perform when the future of the external call completes is limited to being an effect (i.e. persisting zero or more events and optionally replying to the original replyee).

It is only intended to be used on the actor context of the enclosing behavior and not any other actor contexts (indeed the API should probably be modified if possible to only allow the effect to be applied to the self actor). 

As stated earlier the main benefit of having this supported directly by Akka is that the user does not have to define the private callback commands. If it is possible to create an interceptor that handles messages that are not sub types of the targeted behaviors declared command type, this would allow for a user land solution which would not require any Akka modifications. I will investigate that possibility further.

**Comment by patriknw** [comment]
I'm against this. Adding callbacks that are run at arbitrary later point makes the actor difficult to reason about. That is especially true in case of errors and when switching behavior inbetween. The actor should have one entry point for incoming messages (well, two if you count signals).

Reducing boilerplate of having to define the extra commands doesn't motivate the complexity of callback reasoning and additional concepts introduced by this.

**Comment by odd** [comment]
Hi Patrik and thanks for your input. I will close this PR now and hopefully in the future a better method will be found that keeps the amount of boiler plate in check while still avoiding the downsides you mention. Thanks again for taking the time to review it!

**Comment by patriknw** [comment]
Thanks for the suggestion, and for understanding. Keep em coming.
